elementpath/xpath_nodes.py

```diff
@@ -136,480 +161,755 @@ class NamespaceNode(XPathNode):
+    @property
+    def string_value(self) -> str:
+        return ''.join(etree_iter_strings(self.document.getroot()))
 
-        yield from etree_iter_paths(child, child_path)
+    @property
+    def typed_value(self) -> UntypedAtomic:
+        return UntypedAtomic(''.join(etree_iter_strings(self.document.getroot())))
+
+    @property
+    def document_uri(self) -> Optional[str]:
+        try:
+            uri = cast(str, self.document.getroot().attrib[XML_BASE])
+            parts = urlparse(uri)
+        except (KeyError, ValueError):
+            pass
+        else:
+            if parts.scheme and parts.netloc or parts.path.startswith('/'):
+                return uri
+        return None
 
 
-###
-# XPath node test functions
-#
-# XPath has there are seven kinds of nodes:
 #
-#  element, attribute, text, namespace, processing-instruction, comment, document
+# Helper functions and type aliases
+
+def is_schema(obj: Any) -> bool:
+    return hasattr(obj, 'xsd_version') and hasattr(obj, 'maps') and not hasattr(obj, 'type')
+
+
+def is_schema_node(obj: Any) -> bool:
+    return hasattr(obj, 'local_name') and hasattr(obj, 'type') and hasattr(obj, 'name')
+
+
+def is_xpath_node(obj: Any) -> bool:
+    return isinstance(obj, XPathNode) or is_etree_element(obj) or is_etree_document(obj)
+
+
+RootArgType = Union[DocumentProtocol, DocumentNode, ElementProtocol,
+                    XsdElementProtocol, XMLSchemaProtocol, ElementNode]
+
+
 #
-# Element-like objects are used for representing elements and comments,
-# ElementTree-like objects for documents. XPathNode subclasses are used
-# for representing other node types and typed elements/attributes.
-###
-def match_element_node(obj: Any, tag: Optional[str] = None) -> Any:
+# Node trees builders
+def get_node_tree(root: RootArgType, namespaces: Optional[Dict[str, str]] = None) \
+        -> Union[DocumentNode, ElementNode]:
     """
-    Returns `True` if the first argument is an element node matching the tag, `False` otherwise.
-    Raises a ValueError if the argument tag has to be used but it's in a wrong format.
+    Returns a tree of XPath nodes that wrap the provided root tree.
 
-    :param obj: the node to be tested.
-    :param tag: a fully qualified name, a local name or a wildcard. The accepted
-    wildcard formats are '*', '*:*', '*:local-name' and '{namespace}*'.
+    :param root: an Element or an ElementTree or a schema or a schema element.
+    :param namespaces: an optional mapping from prefixes to namespace URIs, \
+    Ignored if root is a lxml etree or a schema structure.
     """
-    if isinstance(obj, TypedElement):
-        obj = obj.elem
-    elif not is_etree_element(obj) or callable(obj.tag):
-        return False
+    if isinstance(root, (DocumentNode, ElementNode)):
+        return root
+    elif is_etree_document(root):
+        if hasattr(root, 'xpath'):
+            return build_lxml_node_tree(cast(DocumentProtocol, root))
```

Comment:
There is an `iselement()` function in both `lxml.etree` and `xml.etree.ElementTree`.

Reply:
I will try to use `iselement()` from `xml.etree.ElementTree`. The method `is_etree_document()` is for checking if it's an ElementTree instance.

Reply:
The `iselement()` of  `xml.etree.ElementTree` checks only the 'tag' attribute. Before using that i have to verify if it can be a problem with some schema components (this was the reason in the past).

Reply:
`iselement()` in `lxml.etree` is different from that. It does a proper type check. Since the elements in ElementTree don't have an `xpath` method, this is probably what you want here, right?

Reply:
Type checking is not an option in this case because the elements could be of of different classes (also schema elements), so an attribute based check is what i want here currently.